ledger/src/blockstore_processor.rs

```diff
@@ -142,237 +143,212 @@ fn get_first_error(
+                    dep_graph.extend(indices.iter().take_while(|l_idx| **l_idx < idx));
+                }
+            }
 
-    tx_batch
+            for read_acc in writelock_accs {
+                if let Some(indices) = indicies_write_locking_account.get(read_acc) {
+                    dep_graph.extend(indices.iter().take_while(|l_idx| **l_idx < idx));
+                }
+            }
+            dep_graph
+        })
+        .collect())
 }
 
-fn execute_batches(
-    bank: &Arc<Bank>,
-    batches: &[TransactionBatch],
-    entry_callback: Option<&ProcessCallback>,
-    transaction_status_sender: Option<&TransactionStatusSender>,
-    replay_vote_sender: Option<&ReplayVoteSender>,
-    timings: &mut ExecuteTimings,
-    cost_capacity_meter: Arc<RwLock<BlockCostCapacityMeter>>,
-) -> Result<()> {
-    let (lock_results, sanitized_txs): (Vec<_>, Vec<_>) = batches
-        .iter()
-        .flat_map(|batch| {
-            batch
-                .lock_results()
-                .iter()
-                .cloned()
-                .zip(batch.sanitized_transactions().to_vec())
-        })
-        .unzip();
-
-    let cost_model = CostModel::new();
-    let mut minimal_tx_cost = u64::MAX;
-    let mut total_cost: u64 = 0;
-    // Allowing collect here, since it also computes the minimal tx cost, and aggregate cost.
-    // These two values are later used for checking if the tx_costs vector needs to be iterated over.
-    #[allow(clippy::needless_collect)]
-    let tx_costs = sanitized_txs
-        .iter()
-        .map(|tx| {
-            let cost = cost_model.calculate_cost(tx).sum();
-            minimal_tx_cost = std::cmp::min(minimal_tx_cost, cost);
-            total_cost = total_cost.saturating_add(cost);
-            cost
-        })
-        .collect::<Vec<_>>();
-
-    let target_batch_count = get_thread_count() as u64;
-
-    let mut tx_batches: Vec<TransactionBatch> = vec![];
-    let rebatched_txs = if total_cost > target_batch_count.saturating_mul(minimal_tx_cost) {
-        let target_batch_cost = total_cost / target_batch_count;
-        let mut batch_cost: u64 = 0;
-        let mut slice_start = 0;
-        tx_costs.into_iter().enumerate().for_each(|(index, cost)| {
-            let next_index = index + 1;
-            batch_cost = batch_cost.saturating_add(cost);
-            if batch_cost >= target_batch_cost || next_index == sanitized_txs.len() {
-                let tx_batch =
-                    rebatch_transactions(&lock_results, bank, &sanitized_txs, slice_start, index);
-                slice_start = next_index;
-                tx_batches.push(tx_batch);
-                batch_cost = 0;
+fn build_batch_indices(tx_dependency_graph: &Vec<HashSet<usize>>) -> Vec<Vec<usize>> {
+    let mut processed_indices = vec![false; tx_dependency_graph.len()];
+    let mut batches = Vec::new();
+
+    let mut batch = Vec::with_capacity(tx_dependency_graph.len());
+
+    while !processed_indices.iter().all(|processed| *processed) {
```

Comment:
```suggestion
    while processed_indices.iter().any(|processed| !*processed) {
```

Reply:
^ I imagine this would be faster because then we just loop until we find the first unprocessed. Rather than looping over every tx everytime.

...or, could just check a counter. We know up-front the number of tx, so just count how many transactions we've processed and check if it equals the length of `processed_indices`

Reply:
nice thanks!

Reply:
from profiling it seems like the extend call w/ hashset is the slowest thing

Reply:
We could pre-allocate the sets with capacity idx - 1 since we know that's the upper bound. It'll use more memory, but if the batches are small it's not a concern and could potentially avoid many reallocations.

Reply:
yeah its unclear to me if its that or the hashing algorithm... will dig into it more and see what i find

Reply:
[FxHasher](https://docs.rs/rustc-hash/1.1.0/rustc_hash/struct.FxHasher.html
) and [FnvHasher](https://docs.rs/fnv/latest/fnv/struct.FnvHasher.html) could both be good alternatives; these are both better suited than default when using small keys (such as our `usize`s). Collision resistance is an issue with these algorithms, need to think a bit more on this. The attack vector would be a malicious leader crafting a block such that transactions are certain indices collide ... need to think on this a little more

> We could pre-allocate the sets with capacity idx - 1 since we know that's the upper bound. It'll use more memory, but if the batches are small it's not a concern and could potentially avoid many reallocations.

If we end up considering this route, direct indexing into a bitfield might be more space & time efficient

Reply:
yeah... i tried fxhash + ahash w/ mixed results. another idea was intelligently trying to figure out how much needs to be expanded since we might be trying to add the same elements a few times.